pkg/engine/internal/scheduler/wire/wire_http2.go

```diff
@@ -0,0 +1,357 @@
+	connAcceptTimeout time.Duration
+}
+
+var _ Listener = (*HTTP2Listener)(nil)
+
+type http2ListenerOpts struct {
+	// ConnAcceptTimeout defines how long to wait for a connection to be accepted via .Accept().
+	// The connection is closed if Accept() is too slow.
+	ConnAcceptTimeout time.Duration
+
+	// MaxPendingConns defines the maximum number of pending connections (which are not Accepted yet).
+	MaxPendingConns uint
+
+	// Logger is used for logging.
+	Logger log.Logger
+}
+
+type HTTP2ListenerOptFunc func(*http2ListenerOpts)
+
+func WithHTTP2ListenerConnAcceptTimeout(timeout time.Duration) HTTP2ListenerOptFunc {
+	return func(o *http2ListenerOpts) {
+		o.ConnAcceptTimeout = timeout
+	}
+}
+
+func WithHTTP2ListenerMaxPendingConns(maxPendingConns uint) HTTP2ListenerOptFunc {
+	return func(o *http2ListenerOpts) {
+		o.MaxPendingConns = maxPendingConns
+	}
+}
+
+func WithHTTP2ListenerLogger(logger log.Logger) HTTP2ListenerOptFunc {
+	return func(o *http2ListenerOpts) {
+		o.Logger = logger
+	}
+}
+
+// NewHTTP2Listener creates a new HTTP/2 listener on the specified address.
+func NewHTTP2Listener(
+	addr net.Addr,
+	protocol FrameProtocol,
+	optFuncs ...HTTP2ListenerOptFunc,
+) *HTTP2Listener {
+	opts := http2ListenerOpts{
+		ConnAcceptTimeout: 1 * time.Second,
+		MaxPendingConns:   10,
+		Logger:            log.NewNopLogger(),
+	}
+	for _, optFunc := range optFuncs {
+		optFunc(&opts)
+	}
+
+	l := &HTTP2Listener{
+		addr:   addr,
+		logger: opts.Logger,
+
+		connCh:            make(chan Conn, opts.MaxPendingConns),
+		closed:            make(chan struct{}),
+		protocol:          protocol,
+		connAcceptTimeout: opts.ConnAcceptTimeout,
+	}
+
+	return l
+}
+
+// HandleStream handles incoming stream connections.
+func (l *HTTP2Listener) HandleStream(w http.ResponseWriter, r *http.Request) {
+	if r.Method != http.MethodPost {
+		http.Error(w, "method not allowed", http.StatusMethodNotAllowed)
+		return
+	}
+
+	flusher, ok := w.(http.Flusher)
+	if !ok {
+		http.Error(w, "streaming not supported", http.StatusInternalServerError)
+		return
+	}
+
+	if r.ProtoMajor != 2 {
+		http.Error(w, "protocol not supported", http.StatusHTTPVersionNotSupported)
+		return
+	}
+
+	w.WriteHeader(http.StatusOK)
+	flusher.Flush()
+
+	conn := newHTTP2Conn(l.Addr(), r.RemoteAddr, r.Body, w, flusher, l.protocol)
+
+	// Try to enqueue the connection without blocking indefinitely
+	select {
+	case <-l.closed:
+		err := conn.Close()
+		if err != nil {
+			level.Error(l.logger).Log("msg", "failed to close connection on listener close", "err", err.Error())
+		}
+		return
+	case l.connCh <- conn:
+		// Connection accepted, wait for it to close
+		<-conn.done
+	case <-time.After(l.connAcceptTimeout):
+		// If Accept() is too slow, close the connection
+		err := conn.Close()
+		if err != nil {
+			level.Error(l.logger).Log("msg", "failed to close not accepted connection", "err", err.Error())
+		}
+
+		http.Error(w, "connection not accepted", http.StatusServiceUnavailable)
```

Comment:
I don't believe this will work because we're already calling `w.WriteHeader(http.StatusOK)` before the select statement.